scilpy/reconst/bingham.py

```diff
@@ -0,0 +1,388 @@
+                                              itertools.repeat(max_lobes),
+                                              itertools.repeat(max_fit_angle)))
+    pool.close()
+    pool.join()
+
+    out = np.concatenate(out, axis=0)
+    out = out.reshape(np.append(shape[:3], max_lobes*NB_PARAMS))
+    return out
+
+
+def _bingham_fit_sh_chunk(args):
+    """
+    Fit Bingham functions on a (N, ncoeffs) chunk taken from a SH field.
+    """
+    sh_chunk = args[0]
+    B_mat = args[1]
+    sphere = args[2]
+    abs_th = args[3]
+    min_sep_angle = args[4]
+    rel_th = args[5]
+    max_lobes = args[6]
+    max_angle = args[7]
+
+    out = np.zeros((len(sh_chunk), max_lobes*NB_PARAMS))
+    for i, sh in enumerate(sh_chunk):
+        odf = sh.dot(B_mat)
+        odf[odf < abs_th] = 0.
+        if (odf > 0.).any():
+            lobes = \
+                _bingham_fit_multi_peaks(odf, sphere, max_angle,
+                                         min_sep_angle, rel_th)
+            for ll in range(min(len(lobes), max_lobes)):
+                lobe = lobes[ll]
+                out[i, ll*NB_PARAMS:(ll+1)*NB_PARAMS] = lobe.get_flatten()
+    return out
+
+
+def _bingham_fit_multi_peaks(odf, sphere, max_angle,
+                             min_sep_angle, rel_th):
+    """
+    Peak extraction followed by Bingham fit for each peak.
+    """
+    peaks, _, _ = peak_directions(odf, sphere,
+                                  relative_peak_threshold=rel_th,
+                                  min_separation_angle=min_sep_angle)
+
+    lobes = []
+    for peak in peaks:
+        peak_fit = _bingham_fit_peak(odf, peak, sphere, max_angle)
+        lobes.append(peak_fit)
+
+    return lobes
+
+
+def _bingham_fit_peak(sf, peak, sphere, max_angle):
+    """
+    Fit bingham function on the lobe aligned with peak.
+    """
+    # abs for twice the number of pts to fit
+    dot_prod = np.abs(sphere.vertices.dot(peak))
+    min_dot = cos(radians(max_angle))
+
+    p = sphere.vertices[dot_prod > min_dot]
+    v = sf[dot_prod > min_dot].reshape((-1, 1))  # (N, 1)
+
+    # test that the peak contains at least 3 non-zero directions
+    if np.count_nonzero(v) < 3:
+        return BinghamDistribution(0, np.zeros(3), np.zeros(3), 0, 0)
+
+    x, y, z = (p[:, 0:1], p[:, 1:2], p[:, 2:])
+
+    # create an orientation matrix to approximate mu0, mu1 and mu2
+    T = np.zeros((3, 3))
+    T[0, 0] = np.sum(x**2 * v)
+    T[1, 1] = np.sum(y**2 * v)
+    T[2, 2] = np.sum(z**2 * v)
+    T[1, 0] = np.sum(x * y * v)
+    T[2, 0] = np.sum(x * z * v)
+    T[2, 1] = np.sum(y * z * v)
+    T[0, 1] = T[1, 0]
+    T[0, 2] = T[2, 0]
+    T[1, 2] = T[2, 1]
+    T = T / np.sum(v)
+
+    eval, evec = np.linalg.eig(T)
+
+    ordered = np.argsort(eval)
+    mu1 = evec[:, ordered[1]].reshape((3, 1))
+    mu2 = evec[:, ordered[0]].reshape((3, 1))
+    f0 = v.max()
+
+    if np.iscomplex(mu1).any() or np.iscomplex(mu2).any():
+        print('uh oh... \n', eval)
```

Comment:
Please do a proper error message.

Reply:
Oops that's old debug code. If the condition is met, it means a Bingham fit was not possible for this voxel. I think it should not crash but instead return 0.

Reply:
Fixed in 1bb08fb